gdnative-core/src/export/macros.rs

```diff
@@ -11,16 +11,43 @@ macro_rules! godot_wrap_method_parameter_count {
     }
 }
 
+#[doc(hidden)]
+#[macro_export]
+macro_rules! godot_wrap_method_if_deref {
+    (true, $ret:expr) => {
+        std::ops::Deref::deref(&$ret)
+    };
+    (false, $ret:expr) => {
+        $ret
+    };
+}
+
+#[doc(hidden)]
+#[macro_export]
+#[deprecated = "This function does not actually pass by reference to the Godot engine. You can clarify by writing #[export(deref_return)]."]
+macro_rules! deprecated_reference_return {
```

Comment:
Yes. (If by "end user" you mean "developer using the library")

However, the method of emit warnings is a terrible hack. This is because there is no way to emit warnings from macros in stable Rust.

Follow these steps to emit warnings.

1. Detect whether reference types are used in derive_methods().  https://github.com/godot-rust/godot-rust/blob/193f7ff8ece2c7b25a44366e23c91c5c9a77581c/gdnative-derive/src/methods.rs#L144-L152
2. Expand the call to the deprecated_reference_return!() macro to user code.  https://github.com/godot-rust/godot-rust/blob/193f7ff8ece2c7b25a44366e23c91c5c9a77581c/gdnative-derive/src/methods.rs#L166

This is hard to understand, so an alternative [proc-macro2-diagnostics](https://crates.io/crates/proc-macro2-diagnostics) crate can be used. (However, an error is emitted by stable Rust)

Reply:
Thanks for explaining! I'm OK with this approach, but could you maybe write 2-3 lines of comments explaining your thoughts behind this design and how it works? That would make future developers less puzzled 🙂

Reply:
Added comments.